core-plugins/src/main/java/io/cdap/plugin/batch/connector/FileConnector.java

```diff
@@ -0,0 +1,144 @@
+import java.nio.file.attribute.PosixFilePermissions;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Set;
+
+/**
+ * File connector to browse flat file on local system
+ */
+@Plugin(type = Connector.PLUGIN_TYPE)
+@Name(FileConnector.NAME)
+@Description("Flat file connector for local file system")
+@Category("Flat File")
+public class FileConnector extends AbstractFileConnector<FileConnector.FileConnectorConfig> {
+  public static final String NAME = "File";
+  static final String FILE_TYPE_KEY = "File Type";
+  static final String SIZE_KEY = "Size";
+  static final String LAST_MODIFIED_KEY = "Last Modified";
+  static final String OWNER_KEY = "Owner";
+  static final String GROUP_KEY = "Group";
+  static final String PERMISSION_KEY = "Permission";
+  private static final String DIRECTORY = "directory";
+  private static final String FILE = "file";
+
+  @Override
+  public BrowseDetail browse(ConnectorContext connectorContext, BrowseRequest request) throws IOException {
+    String path = request.getPath();
+    File file = new File(request.getPath());
+    if (!file.exists()) {
+      throw new IllegalArgumentException(String.format("The given path %s does not exist", path));
+    }
+
+    // if it is not a directory, then it is not browsable, return the path itself
+    if (!file.isDirectory()) {
+      return BrowseDetail.builder().setTotalCount(1).addEntity(generateBrowseEntity(file)).build();
+    }
+
+    // list the files and classify them with file and directory
+    File[] files = file.listFiles();
+    if (files == null) {
+      throw new IOException(String.format("Unable to browse the path %s", path));
+    }
+
+    // sort the files by names
+    Arrays.sort(files);
+    int limit = request.getLimit() == null ? files.length : Math.min(request.getLimit(), files.length);
+    BrowseDetail.Builder builder = BrowseDetail.builder().setTotalCount(files.length);
+    for (int i = 0; i < limit; i++) {
+      // do not browse hidden files
+      if (files[i].isHidden()) {
+        continue;
+      }
+      builder.addEntity(generateBrowseEntity(files[i]));
+    }
+    return builder.build();
+  }
+
+  @Override
+  protected void setConnectorSpec(ConnectorSpecRequest request, ConnectorSpec.Builder builder) {
+    super.setConnectorSpec(request, builder);
+    builder.addRelatedPlugin(new PluginSpec(FileBatchSource.NAME, BatchSource.PLUGIN_TYPE,
+                                            Collections.singletonMap("path", request.getPath())));
+  }
+
+  private BrowseEntity generateBrowseEntity(File file) throws IOException {
+    String path = file.getCanonicalPath();
+    boolean isDirectory = file.isDirectory();
+    BrowseEntity.Builder builder = BrowseEntity.builder(file.getName(), path, isDirectory ? DIRECTORY : FILE);
+    Set<BrowseEntityProperty> properties = new HashSet<>();
+
+    if (isDirectory) {
+      builder.canBrowse(true).canSample(true);
```

Comment:
why can directory be sampled ?

Reply:
File source is frequently setup to read from a directory, in which case it reads all files in the directory.